### DISCUSSION TITLE
Why is static build is not supported?

### DISCUSSION THREAD

**Discussion by tinywrkb** [fish-shell/fish-shell#9207] [Q&A] [answered]
As someone who just started trying fish, but do build it as static PIE binary, I would like to know why this is not supported, and what are the gotchas to be aware of?  
It does look like I'm going to have to deal with fish datadir detection in my config, maybe also sourcing functions and completions if the default `datadir/config.fish` can't deal with it, and very likely patching `datadir/__fish_build_paths.fish`.

**Comment by faho** [comment, marked as answer]
Because it doesn't work in common configurations - glibc's name lookups break, for instance - https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged. This is a real issue, see #7947 - where fish crashes if you enter `~`, because it'll try to look up your system's users.

And it also doesn't *help* much, because you still won't have a single-file relocatable build of fish, you still need a bunch of auxilliary files.

The best solution for that is something like an appimage - see #6475.

So because it breaks stuff and isn't of much use, we declared it unsupported. If you manage to make it work, have fun with it. Just don't file any bugs.

**Comment by tinywrkb** [comment]
Thanks for answering.

> This is a real issue, see https://github.com/fish-shell/fish-shell/issues/7947 - where fish crashes if you enter ~, because it'll try to look up your system's users.

No such issue when building with musl libc 1.2.3 and GCC 12.2.0.  
Tried in updated Arch Linux host environment, Alpine edge Podman container, and Flatpak sandbox.  
The only limitation here is that it doesn't detect systemd-homed users in the host environment.

> And it also doesn't help much, because you still won't have a single-file relocatable build of fish, you still need a bunch of auxilliary files.

Having a single file is not the point, it is more about being able to take all your tools and configs, use them everywhere, while not polluting the environment with conflicting shared libs.  
I can build all the CLI and TUI tools that I usually use as static executables, install them to a custom prefix, and then bind mount this location into every container I want.

> The best solution for that is something like an appimage

I'm not a fan of this solution, it adds an overhead due to SquashFS, and if I remember correctly it needs FUSE to operate, which I find unacceptable, I'm not going to poke holes in my containers for this.  
I really don't see the point of an AppImage if the only issue is detection of fish datadir.

> If you manage to make it work, have fun with it. Just don't file any bugs.

Well, so far it works fine, and should be enough for me. I posted here as I've seen bug reports being locked, and developers didn't seem keen about this, so I wondered why.

**Reply by faho** [reply]
> No such issue when building with musl libc 1.2.3 and GCC 12.2.0.

The point is that it's not worth spending time debugging because many problems will just be down to problems with static linking, so it's also not worth enumerating the cases where it probably works.

I *think* it works on musl, but I have no interest in confirming that enough to support it, and to figure out what to add so it doesn't print the "it's unsupported" error on musl.

>I posted here as I've seen bug reports being locked

We autolock closed issues after a year of inactivity to avoid people adding their new problems onto an old solved bug. If you see something that's been fixed a year ago and you still have similar problems, it's a new issue and should be filed as a new issue, complete with new reproduction information ("what's your OS", "can you reproduce without configuration" - the sort of stuff the issue template asks for).

It's basically a prohibition on "necrobumping".

>developers didn't seem keen about this

See above. We've had a few cases of people reporting weird issues that just boiled down to static linking so we simply tell them it's unsupported to avoid wasting time on debugging something only to figure out it's static linking.

We're a small team of volunteers and so it helps to cut down on support area.

**Reply by tinywrkb** [reply]
Thanks again for the detailed reply. This clarifies more the developers' stand and reasoning on the matter than the current note in the readme.  
It would've been nice if that note said something like:

> There are known issues when statically linking with glibc, and it should be avoided. Statically linking with musl *might* work, but it's unsupported, and don't file bug reports if you do so.
